File: Ruby/string_incrementer.rb
# 5 kyu
# String incrementer

# Your job is to write a function which increments a string, to create a new string.
# If the string already ends with a number, the number should be incremented by 1.
# If the string does not end with a number. the number 1 should be appended to the new string.
# Attention: If the number has leading zeros the amount of digits should be considered.

# split the string by a-z
#   if array.empty?
#     suffix a "1" to the original string
#   else the last char is a number string
#     number.to_i + 1
#   end

def string_increment(string)
  result = string.split(/[a-z]/i)
  new_str = ''
  if string[-1] =~ /[a-z]/i || string == ""
    new_str = string + "1"
  else
    digits_size = result.last.size
    if ((result.last.to_i + 1).to_s).size < digits_size # means 0 doestn't matter
      new_str = string[0, string.size-result.last.to_i.to_s.size]+(result.last.to_i + 1).to_s
    else
      new_str = string[0, string.size - digits_size] + (result.last.to_i + 1).to_s
    end
  end
  new_str
end

p string_increment("foo") == "foo1"
p string_increment("foobar23") == "foobar24"
p string_increment("foo0042") == "foo0043"
p string_increment("foo9") == "foo10"
p string_increment("foo099") == "foo100"
p string_increment("") == "1"

File: JavaScript/string_incrementer.js
// # 5 kyu
// # String incrementer

// # Your job is to write a function which increments a string, to create a new string.
// # If the string already ends with a number, the number should be incremented by 1.
// # If the string does not end with a number. the number 1 should be appended to the new string.
// # Attention: If the number has leading zeros the amount of digits should be considered.

// input:
//   - string
// output:
//   - a new string, ending with a number
// rules:
//   - if the string is not ended with 0-9, add a "1" to it

// algo:
//   - check if the string containing any numbers
//   - if no, concat a "1" to the end of it, return it
//   - if yes, convert the string to array
//     - filter out those string that are numbers
//     - concat them togther
//     - convert them into numbers
//     - do the math, + 1
//     - convert the number back to string, 
//     - combine the letter part with this number part
//   - return the final result

function stringIncrement(string) {
  if (!string.match(/[0-9]/g)) return string + "1";
  
  let allChars = string.split('');
  let allNums = allChars.filter(char => char.match(/[0-9]/g)).join(''); // "0042", "099"
  let number = Number(allNums) + 1; // 43, 100

  let numPart = String(number); // tricky part
  if (allNums.length !== String(number).length) {
    numPart = String(number).padStart(allNums.length, "0");
  }
  
  let letterPart = (allChars.filter(char => char.match(/[a-z]/ig))).join('');
  return letterPart + numPart;
}

console.log(stringIncrement("foo")) // "foo1"
console.log(stringIncrement("foobar23")) // "foobar24"
console.log(stringIncrement("foo0042")) // "foo0043"
console.log(stringIncrement("foo9")) // "foo10"
console.log(stringIncrement("foo099")) // "foo100"
console.log(stringIncrement("")) // "1"
